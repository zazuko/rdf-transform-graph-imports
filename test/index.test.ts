import * as url from 'url'
import * as http from 'http'
import chai, { expect } from 'chai'
import type { Dataset } from '@zazuko/env/lib/Dataset.js'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { Readable } from 'readable-stream'
import transform from '../index.js'
import rdf from './env.js'
import { start } from './server/index.js'

chai.use(jestSnapshotPlugin())

describe('rdf-merge-stream', () => {
  let server: http.Server

  before(() => {
    server = start().listen(6666)
  })

  after(() => {
    server.close()
  })

  it('merges file stream by relative path', async () => {
    // given
    const path = new URL('./resources/import-relative.ttl', import.meta.url)
    const root = rdf.fromFile(url.fileURLToPath(path.toString()))

    // when
    const merged = await rdf.dataset().import(root.pipe(transform(rdf, {
      basePath: path,
    })))

    // then
    expect(await turtle(merged)).toMatchSnapshot()
  })

  it('imports remote resources from files', async () => {
    // given
    const path = new URL('./resources/import-remote.ttl', import.meta.url)
    const root = rdf.fromFile(url.fileURLToPath(path.toString()))

    // when
    const merged = await rdf.dataset().import(root.pipe(transform(rdf, {
      basePath: path,
    })))

    // then
    expect(await turtle(merged)).toMatchSnapshot()
  })

  it('imports remote resources from http', async () => {
    // given
    const response = await rdf.fetch('http://localhost:6666/shape')
    const root = await response.quadStream() as unknown as Readable

    // when
    const merged = await rdf.dataset().import(root.pipe(transform(rdf)))

    // then
    expect(await turtle(merged)).toMatchSnapshot()
  })
})

async function turtle(merged: Dataset) {
  return merged.serialize({ format: 'text/turtle', prefixes: ['sh', 'schema'] })
}
