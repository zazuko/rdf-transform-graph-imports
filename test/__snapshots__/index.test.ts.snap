// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rdf-merge-stream imports remote resources from files 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix schema: <http://schema.org/> .

<http://example.org/required-property> sh:minCount 1 ;
	sh:maxCount 1 .

<http://localhost:6666/property/identifier> sh:minLength 3 ;
	sh:and (
		<http://example.org/required-property>
	) ;
	sh:path schema:identifier .

<> a sh:NodeShape ;
	sh:property <http://localhost:6666/property/identifier> .

"
`;

exports[`rdf-merge-stream imports remote resources from http 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix schema: <http://schema.org/> .

<http://example.org/required-property> sh:minCount 1 ;
	sh:maxCount 1 .

<http://localhost:6666/property/identifier> sh:minLength 3 ;
	sh:and (
		<http://example.org/required-property>
	) ;
	sh:path schema:identifier .

<http://localhost:6666/shape> a sh:NodeShape ;
	sh:property <http://localhost:6666/property/identifier> .

"
`;

exports[`rdf-merge-stream merges file stream by relative path 1`] = `
"@prefix sh: <http://www.w3.org/ns/shacl#> .
@prefix schema: <http://schema.org/> .

<http://example.org/required-property> sh:minCount 1 ;
	sh:maxCount 1 .

<#property/identifier> sh:minLength 3 ;
	sh:and (
		<http://example.org/required-property>
	) ;
	sh:path schema:identifier .

<> a sh:NodeShape ;
	sh:property <#property/identifier> .

"
`;
